File: scripts/update-changelog.js
#!/usr/bin/env node
/**
 * Reads a commit message and inserts a bullet into CHANGELOG.md under the
 * correct date + section. Called by scripts/hooks/commit-msg.
 *
 * Usage: node update-changelog.js <commit-msg-file> <changelog-path>
 */

const fs = require('fs')

const msgFile      = process.argv[2]
const changelogPath = process.argv[3]

if (!msgFile || !changelogPath) {
  process.stderr.write('Usage: update-changelog.js <msg-file> <changelog-path>\n')
  process.exit(1)
}

const msg = fs.readFileSync(msgFile, 'utf8').split('\n')[0].trim()

const SECTION_MAP = {
  'FEATURE':     'Features',
  'ENHANCEMENT': 'Enhancements',
  'STYLE':       'Enhancements',
  'BUG':         'Bug Fixes',
  'REFACTOR':    'Refactors',
}

const match = msg.match(/^(FEATURE|ENHANCEMENT|STYLE|BUG|REFACTOR):\s*(.+)/)
if (!match) process.exit(0) // CHORE: and others — skip

const [, prefix, text] = match
const section = SECTION_MAP[prefix]
const entry   = `- ${text.trim()}`
const today   = new Date().toISOString().slice(0, 10) // YYYY-MM-DD

// Read existing changelog or start fresh
let content = '# Changelog\n'
try { content = fs.readFileSync(changelogPath, 'utf8') } catch {}

const lines  = content.split('\n')
const result = []
let i = 0

const dateLine     = `## ${today}`
const sectionLine  = `### ${section}`

let dateFound    = false
let sectionFound = false
let entryPlaced  = false

while (i < lines.length) {
  const line = lines[i]

  // ── Found today's date block ──────────────────────────────────────────
  if (!dateFound && line.trim() === dateLine) {
    dateFound = true
    result.push(line)
    i++
    continue
  }

  if (dateFound && !entryPlaced) {
    // Inside today's block — look for the right section
    if (line.trim() === sectionLine) {
      sectionFound = true
      result.push(line)
      i++
      // Insert entry right after the section header
      result.push(entry)
      entryPlaced = true
      continue
    }

    // Hit the next date block before finding our section — insert section before it
    if (line.startsWith('## ') && line.trim() !== dateLine) {
      result.push(sectionLine)
      result.push(entry)
      result.push('')
      entryPlaced = true
      // fall through to push this line normally
    }
  }

  result.push(line)
  i++
}

// ── Today's date block didn't exist yet — insert at top after "# Changelog" ──
if (!dateFound) {
  const headerIdx = result.findIndex(l => l.startsWith('# '))
  const insertAt  = headerIdx >= 0 ? headerIdx + 1 : 0
  result.splice(insertAt, 0, '', dateLine, '', sectionLine, entry, '')
  entryPlaced = true
}

// ── Date found but section never closed and no next date block ───────────────
if (dateFound && !entryPlaced) {
  result.push(sectionLine)
  result.push(entry)
  result.push('')
}

fs.writeFileSync(changelogPath, result.join('\n'), 'utf8')
process.stdout.write(`changelog: [${today}] ${section} — ${entry}\n`)

File: scripts/update-changelog.go
package main

// update-changelog: reads a commit-msg file and inserts a bullet into
// CHANGELOG.md under the correct date + section.
//
// Usage: go run scripts/update-changelog.go <commit-msg-file> <changelog-path>

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

var sectionMap = map[string]string{
	"FEATURE":     "Features",
	"ENHANCEMENT": "Enhancements",
	"STYLE":       "Enhancements",
	"BUG":         "Bug Fixes",
	"REFACTOR":    "Refactors",
}

func main() {
	if len(os.Args) < 3 {
		fmt.Fprintln(os.Stderr, "usage: update-changelog <commit-msg-file> <changelog-path>")
		os.Exit(1)
	}
	msgFile := os.Args[1]
	changelogPath := os.Args[2]

	msg := firstLine(msgFile)
	section, body := classify(msg)
	if section == "" {
		return // not a tracked prefix — exit 0
	}

	today := time.Now().Format("2006-01-02")
	entry := "- " + body

	updated := insertEntry(readChangelog(changelogPath), today, section, entry)
	if err := os.WriteFile(changelogPath, []byte(updated), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "write changelog: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("changelog: [%s] %s — %s\n", today, section, entry)
}

func firstLine(path string) string {
	f, err := os.Open(path)
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	if sc.Scan() {
		return strings.TrimSpace(sc.Text())
	}
	return ""
}

func classify(msg string) (section, body string) {
	for prefix, sec := range sectionMap {
		tag := prefix + ": "
		if strings.HasPrefix(msg, tag) {
			return sec, strings.TrimSpace(strings.TrimPrefix(msg, tag))
		}
	}
	return "", ""
}

func readChangelog(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return "# Changelog\n\nAll notable changes to Velocity are listed here. Most recent entries appear first.\n\n---\n"
	}
	// normalise CRLF → LF so index arithmetic works on Windows
	return strings.ReplaceAll(string(data), "\r\n", "\n")
}

// insertEntry places entry under section in today's date block,
// creating the block and/or section when absent.
func insertEntry(content, today, section, entry string) string {
	lines := strings.Split(content, "\n")
	dateLine := "## " + today
	sectionLine := "### " + section

	dateIdx := indexOfLine(lines, dateLine)

	if dateIdx == -1 {
		// No block for today — insert right after the first "---" separator
		insertAt := len(lines)
		for i, l := range lines {
			if strings.TrimSpace(l) == "---" {
				insertAt = i + 1
				break
			}
		}
		block := []string{"", dateLine, "", sectionLine, entry, "", "---"}
		return join(splice(lines, insertAt, block))
	}

	// Find where today's block ends (next ## heading or EOF)
	blockEnd := len(lines)
	for i := dateIdx + 1; i < len(lines); i++ {
		if strings.HasPrefix(lines[i], "## ") {
			blockEnd = i
			break
		}
	}

	// Look for the target section inside today's block
	secIdx := -1
	for i := dateIdx + 1; i < blockEnd; i++ {
		if strings.TrimSpace(lines[i]) == sectionLine {
			secIdx = i
			break
		}
	}

	if secIdx != -1 {
		// Section exists — insert entry right after its header
		return join(splice(lines, secIdx+1, []string{entry}))
	}

	// Section missing — add it before the block's trailing separator
	insertAt := blockEnd
	for insertAt > dateIdx+1 && strings.TrimSpace(lines[insertAt-1]) == "" {
		insertAt--
	}
	if insertAt > 0 && strings.TrimSpace(lines[insertAt-1]) == "---" {
		insertAt--
	}
	return join(splice(lines, insertAt, []string{"", sectionLine, entry}))
}

func indexOfLine(lines []string, target string) int {
	for i, l := range lines {
		if strings.TrimSpace(l) == target {
			return i
		}
	}
	return -1
}

func splice(s []string, at int, elems []string) []string {
	out := make([]string, 0, len(s)+len(elems))
	out = append(out, s[:at]...)
	out = append(out, elems...)
	out = append(out, s[at:]...)
	return out
}

func join(lines []string) string { return strings.Join(lines, "\n") }
